providers/implementations/keymgmt/mlkem_kmgmt-c.inc

```diff
@@ -0,0 +1,568 @@
+    memcpy(pubenc(key), pubenc, PUBKEY_BYTES);
+    key->keytype = MLKEM_KEY_TYPE_PUBLIC;
+    return 1;
+}
+
+static int mlkem_import(void *vkey, int selection, const OSSL_PARAM params[])
+{
+    MLKEM_KEY *key = vkey;
+    int include_private;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+
+    include_private = selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY ? 1 : 0;
+    return ossl_mlkem_key_fromdata(key, params, include_private);
+}
+
+static const OSSL_PARAM *mlkem_gettable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_SECURITY_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_MAX_SIZE, NULL),
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PRIV_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+/*
+ * It is assumed the key is guaranteed non-NULL here, and is from this provider
+ */
+static int mlkem_get_params(void *vkey, OSSL_PARAM params[])
+{
+    MLKEM_KEY *key = vkey;
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, BITS))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_SECURITY_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, SECBITS))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_MAX_SIZE);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, CIPHERTEXT_BYTES))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p != NULL && key->keytype != 0
+        && !OSSL_PARAM_set_octet_string(p, pubenc(key), PUBKEY_BYTES))
+        return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PRIV_KEY);
+    if (p != NULL && key->keytype == MLKEM_KEY_TYPE_PRIVATE
+        && !OSSL_PARAM_set_octet_string(p, prvenc(key), PRVKEY_BYTES))
+        return 0;
+
+    return 1;
+}
+
+static const OSSL_PARAM *mlkem_settable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+static int mlkem_set_params(void *vkey, const OSSL_PARAM params[])
+{
+    MLKEM_KEY *key = vkey;
+    const OSSL_PARAM *p;
+    const void *pubenc = NULL;
+    size_t publen = 0;
+
+    if (params == NULL)
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
+        return 1;
+    if (OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen) != 1)
+        return 0;
+
+    /* This resets the key to public on success, and clears it on failure */
+    if (key->keytype == MLKEM_KEY_TYPE_PRIVATE) {
+        cleanse(key);
+        key->keytype = 0;
+    }
```

Comment:
This is currently broken, because enforcing non-mutation of private to public breaks `evp_extra_test`. Either the test is doing the wrong thing, or we should be returning key material as "parameters", or we need to allow the mutation.

Reply:
This should be fixed, after making "dup" more *selective* (not copy the keys when not requested).